streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStreamImpl.java

```diff
@@ -0,0 +1,196 @@
+                return aggregate + 1;
+            }
+        }, windows, Serdes.Long());
+    }
+
+
+    private <W extends Window, T> StateStoreSupplier windowedStore(final Serde<T> aggValSerde,
+                                                                   final Windows<W> windows) {
+        return storeFactory(aggValSerde, windows.name())
+            .windowed(windows.maintainMs(), windows.segments, false)
+            .build();
+
+    }
+
+    private <T> Stores.PersistentKeyValueFactory<K, T> storeFactory(final Serde<T> aggValueSerde,
+                                                                    final String name) {
+        return Stores.create(name)
+            .withKeys(keySerde)
+            .withValues(aggValueSerde)
+            .persistent();
+
+    }
+
+    private <T> KTable<K, T> doAggregate(
+        final KStreamAggProcessorSupplier<K, ?, V, T> aggregateSupplier,
+        final String functionName,
+        final StateStoreSupplier storeSupplier) {
+
+        final String aggFunctionName = topology.newName(functionName);
+
+        final String sourceName = repartitionIfRequired();
+
+        topology.addProcessor(aggFunctionName, aggregateSupplier, sourceName);
+        topology.addStateStore(storeSupplier, aggFunctionName);
+
+        return new KTableImpl<>(topology,
+                                aggFunctionName,
+                                aggregateSupplier,
+                                sourceName.equals(this.name) ? sourceNodes
+                                                             : Collections.singleton(sourceName));
+    }
+
+    /**
+     * If repartitioning is required then add a new co-partitioned internal topic and configure the
+     * source & sink.
+     *
+     * @return the new sourceName if repartitioned. Otherwise the name of this stream
+     */
+    private String repartitionIfRequired() {
+        if (!repartitionRequired) {
+            return this.name;
+        }
+
+        final Serializer<K> keySerializer = keySerde == null ? null : keySerde.serializer();
+        final Deserializer<K> keyDeserializer = keySerde == null ? null : keySerde.deserializer();
+        final Serializer<V> valueSerializer = valSerde == null ? null : valSerde.serializer();
+        final Deserializer<V> valueDeserializer = valSerde == null ? null : valSerde.deserializer();
+
+        final String sinkName = topology.newName(KStreamImpl.SINK_NAME);
+        final String topic = this.name + KStreamImpl.REPARTITION_TOPIC_SUFFIX;
+        final Set<String> allSourceNodes = new HashSet<>(sourceNodes);
+        final String sourceName = topology.newName(KStreamImpl.SOURCE_NAME);
+        allSourceNodes.add(sourceName);
+        topology.addInternalTopic(topic);
+        topology.addSink(sinkName, topic, keySerializer, valueSerializer, this.name);
+        topology.addSource(sourceName, keyDeserializer, valueDeserializer, topic);
+        topology.copartitionSources(allSourceNodes);
```

Comment:
My concern of calling `topology.copartitionSources(allSourceNodes);` would require ALL source node topics, as well as this internal topic to be co-partitioned. Imagine you have
```
stream = builder.stream("a*")
```

where the regex maps to two topics: `ab` and `ac`, with `m` and `n` partitions. Today we allow `m` and `n` to be different, and by default we create number of tasks equal to the larger of the two; but once the stream is grouped then suddenly `m` and `n` must be equal.

And in fact, I think currently the auto created topics should not be auto created with default (i.e. 1) number of partitions, but with the number of tasks that is writing to this topic, hence in the above case `max(m, n)`. Did you observe it is actually not the case without calling `copartition`?